### DISCUSSION TITLE
Where to start using rust-analyzer programmatically?

### DISCUSSION THREAD

**Discussion by sanmai-NL** [rust-lang/rust-analyzer#14774] [Q&A] [answered]
I'd like to e.g., list all structs that implement a certain trait within a crate programmatically for new tooling I'm developing, separate from an IDE. Ideally I'd use the LSP, but I can use the Rust API. Where to start?

**Comment by lnicola** [comment]
Before trying that, you should also make sure that the JSON output of rustdoc and our LSIF index don't provide enough information.

**Reply by sanmai-NL** [reply]
Are there samples of this information somewhere online for a quick look?

**Reply by lnicola** [reply, marked as answer]
Don't think so, but try `cargo +nightly rustdoc --package=$package -- -Z unstable-options --output-format=json` and `rust-analyzer lsif .` in a project.

**Reply by sanmai-NL** [reply]
@lnicola Did you mean `cargo doc --message-format=json`? That doesn't produce this information (e.g., about traits) on stdout. The `--json` option is unavaible here with Cargo 1.66.0.

**Reply by sanmai-NL** [reply]
Navigating the LSIF file, I don't understand how this encodes which structs there are and which traits they implement.

**Reply by lnicola** [reply]
I was thinking of https://rust-lang.github.io/rfcs/2963-rustdoc-json.html, but I'm not sure how to use it.

**Reply by sanmai-NL** [reply]
The right command for a workspace is:

```sh
cargo +nightly rustdoc --package=$package -- -Z unstable-options --output-format=json
```

The JSON file is saved under `target/doc/`. Indeed, this JSON data seems informative enough (at least for further study).

**Reply by sanmai-NL** [reply]
> Don't think so, but try `cargo doc --json` and `rust-analyzer lsif .` in a project.

Can you update your answer with the correct Cargo command, for future reference?

**Reply by lnicola** [reply]
Done. Thanks for tracking it down, I wasn't near a computer today.
